File: PA3/List.java
//-----------------------------------------------------------------------------
//	Joshua Cheung
//  jotcheun
//  PA3
//  List.java
//-----------------------------------------------------------------------------

public class List
{
	private class Node
	{
		public Object data;
		public Node prev;
		public Node next;

		public Node(Object data)
		{
			this.data = data;
			prev = null;
			next = null;
		}
	}

	private int numitems = 0;
	private Node front;
	private Node back;
	private Node cursor; 
	private int index;

	public List()
	{
		front = back = null;
		cursor = null;
		numitems = 0;
		index = 0;
	}

	public int length()
	{
		return numitems;
	}

	// If cursor is defined, returns the index of the cursor element, 
	// otherwise returns -1.
	int index()
	{  
		if (cursor != null)
		{
			return index; 
		}
		return -1;
	}

	// Returns front element. Pre: length()>0
	Object front()
	{ 
		if (isEmpty())
		{
			throw new RuntimeException("List Error: Front() called on empty List");
		}
		return front.data;
	}

	// Returns back element. Pre: length()>0
	Object back() 
	{ 
		if (isEmpty())
		{
			throw new RuntimeException("List Error: back() called on empty List");
		}
		return back.data;
	}

	// Returns cursor element. Pre: length()>0, index()>=0
	Object get()
	{ 
		if (isEmpty() || index() < 0)
		{
			throw new RuntimeException("List Error: get() called on empty List");
		}
		return cursor.data;
	}

	// Returns true if this List and L are the same integer 
	// sequence. The cursor is ignored in both lists.
	
	boolean equals(List l)
	{   
		boolean equal = false; 
		Node n, m;
		n = this.front;
		m = l.front;
		equal = (length() == l.length());
		while(equal && n!= null)
		{
			equal = (n.data.equals(m.data));
			m = m.next;
			n = n.next;
		}
		return equal;
	}

	boolean isEmpty()
	{
		return numitems == 0;
	}


//---------------------------------------------------------------------------------------------------------------------------------------
//														Manipulation procedures
//---------------------------------------------------------------------------------------------------------------------------------------

	// Resets this List to its original empty state.
	void clear()
	{	 
		front = back = null;
		numitems = 0;
		cursor = null;
		index = 0;
	}
	
	// If List is non-empty, places the cursor under the front element, otherwise does nothing
	void moveFront()
	{   
		if (isEmpty())
		{
			throw new RuntimeException("List Error: moveFront() called on empty List");
		}
		else 
		{
			cursor = front;
			index = 0;
		}
	}

	// If List is non-empty, places the cursor under the back element, otherwise does nothing
	void moveBack()
	{    
		if (isEmpty())
		{
			throw new RuntimeException("List Error: moveback() called on empty List");
		}
		else
		{
			cursor = back;
			index = length()-1;
		}
	}

	// If cursor is defined and not at front, moves cursor one step toward front of this List, if cursor is defined and at front,
	// cursor becomes undefined, if cursor is undefined does nothing.
	void movePrev()
	{     
		if (cursor == front)
		{
			index = -1;
			cursor = null;
		}
		else if (cursor != null)
		{
			cursor = cursor.prev;
			index--;
		}
	}

	// If cursor is defined and not at front, moves cursor one step toward back of this List, if cursor is defined and at back, 
	// cursor becomes undefined, if cursor is undefined does nothing.
	void moveNext()
	{  
		if (cursor == back)
		{
			index = -1;
			cursor = null;
		}
		else 
		{
			cursor = cursor.next;
			index++;
		}
	}

	// Insert new element into this List. If List is non-empty, insertion takes place before front element.
	void prepend(Object data)
	{		
		Node n = new Node(data);
		if (isEmpty())
		{
			front = back = n;
			index++;
			numitems++;
		}
		else
		{
			n.next = front;
			front.prev = n;
			front = n;
			index++;
			numitems++;
		}
	}

	// Insert new element into this List. If List is non-empty, insertion takes place after back element.
	void append(Object data) 
	{		
		Node n = new Node(data);
		if (isEmpty())
		{
			front = back = n;
			numitems++;
		}
		else
		{
			back.next = n;
			n.prev = back;
			back = n;
			numitems++;
		}
	}

	// Insert new element before cursor. Pre: length() > 0, index() >= 0
	void insertBefore(Object data)
	{	
		Node n = new Node(data);
		if (isEmpty() || index() < 0)
		{
			front = back = n;
			index++;
			numitems++;
		} 
		else 
		{
			if (cursor.prev == null)
			{
				prepend(data);
			} 
			else 
			{
				Node temp = cursor.prev;
				temp.next = n;
				n.next = cursor;
				cursor.prev = n;
				n.prev = temp;
				numitems++;
				index++;
			}
		}
	}

	// Insert new element after cursor. Pre: length() > 0, index() >= 0
	void insertAfter(Object data)
	{		
		Node n = new Node(data);
		
		if (isEmpty() || index() < 0)
		{
			front = back = n;
			numitems++;
		}
		else
		{
			if (cursor.next == null)
			{
				append(data);
			}
			else
			{
				Node temp = cursor.next;
				cursor.next = n;
				temp.prev = n;
				n.next = temp;
				n.prev = cursor;
				numitems++;
			}
		}	
	}

	// Deletes the front element. Pre: length()>0
	void deleteFront()
	{     
		if (isEmpty())
		{
			throw new RuntimeException("List Error: deleteFront() called on empty List");
		} 
		else if (front == back)
		{
			cursor = front = back = null;
		}
		else
		{
			if (cursor == front)
			{
				cursor = null;
			}
			front = front.next;
			front.prev = null;
		}
		numitems--;
		index--;
	}

	// Deletes the back element. Pre: length()>0
	void deleteBack()
	{		
		if (isEmpty())
		{
			throw new RuntimeException("List Error: deleteBack() called on empty List");	
		} 
		else if (front == back)
		{
			cursor = front = back = null;
		} 
		else 
		{
			if (cursor == back)
			{
				cursor = null;
			}
			back = back.prev;
			back.next = null;
		}
		numitems--;
	}

	// Deletes cursor element, making cursor undefined. Pre: length() > 0, index >= 0
	void delete()
	{      
		if (isEmpty() || index() < 0)
		{
			throw new RuntimeException("List Error: delete() called on empty List");
		} 
		else if (cursor == front && cursor == back)
		{
			cursor = front = back = null;
			numitems--;
		} 
		else if (cursor == front)
		{
			deleteFront();
		} 
		else if (cursor == back)
		{
			deleteBack();		
		} 
		else if (cursor.prev == null && cursor.next == null) 
		{
			front = back = null;
			index--;
			numitems--;
		} 
		else
		{
			cursor.prev.next = cursor.next;
			cursor.next.prev = cursor.prev;
			cursor = null;
			index--;
			numitems--;
		}
	}
	
	// Overrides Object's toString method. Returns a String representation of this List
	// consisting of a space separated sequence of integers, with front on prev.
	public String toString()
	{   
		StringBuffer sb = new StringBuffer();
      	Node n = front;
     	while (n != null )
     	{
     		sb.append(n.data);
     		if (n.next != null)
     		{
				sb.append(" ");
     		}	
        	n = n.next;
     	}
    	return new String(sb);
	}

	 List copy()
	 {
		List l = new List();

		if (!isEmpty())
		{
			for (Node n = front; n!=null; n = n.next)
			{
				l.append(n.data);
			}
		}
		return l;
	}
	
	List concat(List L)
	{ 
		List C = null;
		List D = null;

		if (L.isEmpty())
		{
			return this;
		}
		else if (this.isEmpty())
		{
			return L;
		}			
		else if (L.isEmpty() && this.isEmpty())
		{
			throw new RuntimeException("List Error: concat() called on empty List");
		}
		else
		{
			C = this.copy();
			D = L.copy();
			System.out.println(C);
			C.back.next = D.front;
			D.front.prev = C.back;
		}		
		C.numitems = C.numitems + L.numitems;
		return C;
	}
}

File: PA5/List.c
//===========================================================================
//Joshua Cheung
//Jotcheun
//PA5
//List.c
//===========================================================================

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include "List.h"

#define TRUE 1
#define FALSE 0

//-------------------------------------------------
//Private type
//-------------------------------------------------

typedef struct NodeObj
{
  	int data;
  	struct NodeObj* prev;
  	struct NodeObj* next;
} NodeObj;

typedef unsigned short bool;


typedef NodeObj* Node;

typedef struct ListObj
{
	Node front;
	Node back;
	Node cursor; 
	int index;
	int numitems;
} ListObj;



Node newNode(int node_Data) 
{
  	Node N = malloc(sizeof(NodeObj));
  	assert(N!=NULL);
  	N->data = node_Data;
  	N->prev = NULL;
  	N->next = NULL;
  	return(N);
}

void freeNode(Node* pN) 
{
	if (pN != NULL && *pN != NULL) 
	{
		free(*pN);
		*pN = NULL;
	}
}

bool isEmpty(List L)
{
	return L->numitems == 0;
}

//-------------------------------------------------
// Constructors-Destructors
//-------------------------------------------------

List newList(void)
{
  	List L = malloc(sizeof(ListObj));
  	assert(L!=NULL);
  	L->front = L->back = L->cursor = NULL;
  	L->numitems = 0;
  	L->index = -1;
  	return L;
}

void clear(List L)
{
	if (L != NULL) 
  	{
  		while (length(L) > 0) 
  		{
			deleteBack(L);
		}
	}
	L->index = -1;

}


void freeList (List* pL)
{
	clear(*pL);
	free(*pL);
	*pL = NULL;
}

// Access functions -----------------------------------------------------------

int length(List L) 
{
	if (L == NULL) 
	{
		printf("List Error: calling length() on NULL List reference\n");
		exit(1);
	}
  	return (L->numitems);
}

int index(List L) 
{  
	if (L->cursor != NULL)
	{
		return (L->index); 
	}
	return -1;
}

int front(List L)
{
	if (isEmpty(L))
	{
		fprintf(stderr, "List Error: Front() called on empty List\n");
		exit(EXIT_FAILURE);
	}
	return L->front->data;
}

int back(List L)
{
	if (isEmpty(L))
	{
		fprintf(stderr, "List Error: back() called on empty List\n");
		exit(EXIT_FAILURE);
	}
	return L->back->data;
}

int get(List L)
{
	if (isEmpty(L) || index(L) < 0)
	{
		fprintf(stderr, "List Error: get() called on empty List\n");
		exit(EXIT_FAILURE);
	}
	return L->cursor->data;
}

int equals(List A, List B)
{
	int equal = 0; 
	Node n, m;
	n = A->front;
	m = B->front;
	equal = (length(A) == length(B));

	while (equal && n!= NULL)
	{
		equal = (n->data == m->data);
		m = m->next;
		n = n->next;
	}
	return equal;
}

// Manipulation procedures ----------------------------------------------------

void moveFront(List L)
{
	if (L == NULL)
  	{
    	fprintf(stderr, "List Error: moveFront() called on empty List\n");
    	exit(EXIT_FAILURE);
  	}
  	else 
	{
		L->cursor = L->front;
		L->index = 0;
	}

}

void moveBack(List L)
{
	if (L == NULL) 
	{
		printf("List Error: calling moveBack() on NULL List reference\n");
		exit(1);
	}
  	else 
	{
		L->cursor = L->back;
		L->index = length(L) - 1;
	}
}

void movePrev(List L)
{
	if (L == NULL) 
	{
		printf("List Error: calling movePrev() on NULL List reference\n");
		exit(1);
	}
	if (L->cursor == L->front)
	{
		L->index = -1;
		L->cursor = NULL;
	}
	else if (L->cursor != NULL)
	{
		L->cursor = L->cursor->prev;
		L->index--;
	}
}

void moveNext(List L)
{
	if (L == NULL) 
	{
		printf("List Error: calling moveNext() on NULL List reference\n");
		exit(1);
	}
	if (L->cursor == L->back)
	{
		L->index = -1;
		L->cursor = NULL;
	}
	else 
	{
		L->cursor = L->cursor->next;
		L->index++;
	}
}

void prepend(List L, int data)
{
	Node n = newNode(data);

	if (L == NULL) 
	{
		printf("List Error: calling prepend() on NULL List reference\n");
		exit(1);
	}
	if (isEmpty(L))
	{
		L->front = L->back = n;
	}
	else
	{
		n->next = L->front;
		L->front->prev = n;
		L->front = n;
	}
	L->index++;
	L->numitems++;
}

void append(List L, int data)
{
	if (L == NULL) 
	{
		printf("List Error: calling append() on NULL List reference\n");
		exit(1);
	}
	if (L->front == NULL)
	{
		L->front = L->back = newNode(data);
	}
	else
	{
		Node n = newNode(data);
		L->back->next = n;
		n->prev = L->back;
		L->back = n;
	}
	L->numitems++;
}

void insertBefore(List L, int data)
{
	Node n = newNode(data);

	if (L == NULL) 
	{
		printf("List Error: calling insertBefore() on NULL List reference\n");
		exit(1);
	}
	if (isEmpty(L) || index(L) < 0)
	{
		L->front = L->back = n;
		L->index++;
		L->numitems++;
		n = NULL;
	} 
	else 
	{
		if (L->cursor->prev == NULL)
		{
			free(n);
			prepend(L, data);
		} 
		else 
		{
			Node temp = L->cursor->prev;
			temp->next = n;
			n->next = L->cursor;
			L->cursor->prev = n;
			n->prev = temp;
			L->numitems++;
			L->index++;
		}
	}
}

void insertAfter(List L, int data)
{
	Node n = newNode(data);
	
	if (L == NULL) 
	{
		printf("List Error: calling insertAfter() on NULL List reference\n");
		exit(1);
	}
	if (isEmpty(L) || index(L) < 0)
	{
		L->front = L->back = n;
		L->numitems++;
	}
	else
	{
		if (L->cursor->next == NULL)
		{
			free(n);
			append(L, data);
		}
		else
		{
			Node temp = L->cursor->next;
			L->cursor->next = n;
			temp->prev = n;
			n->next = temp;
			n->prev = L->cursor;
			L->numitems++;
		}
	}	
}

void deleteFront(List L)
{
	if (L == NULL) 
	{
		printf("List Error: calling deleteFront() on NULL List reference\n");
		exit(1);
	}
	else if (L->front == L->back)
	{
		L->cursor = L->front = L->back = NULL;
	}
	else
	{
		if (L->cursor == L->front)
		{
			L->cursor = NULL;
		}
		L->front = L->front->next;
		freeNode(&L->front->prev);
		L->front->prev = NULL;
	}
	L->numitems--;
	L->index--;
	
}

void deleteBack(List L)
{
	if (L == NULL || length(L) == 0) 
	{
		printf("List Error: calling deleteBack() on NULL List reference\n");
		exit(1);
	}
	else if (L->front == L->back)
	{
		freeNode(&L->front);
		L->cursor = L->front = L->back = NULL;
	} 
	else 
	{
		if (L->cursor == L->back)
		{
			L->cursor = NULL;
		}
		L->back = L->back->prev;
		freeNode(&L->back->next);
		L->back->next = NULL;

	}
	L->numitems--;
}

void delete(List L)
{
	if (L == NULL) 
	{
		printf("List Error: calling delete() on NULL List reference\n");
		exit(1);
	}
	else if (L->cursor == L->front && L->cursor == L->back)
	{
		L->cursor = L->front = L->back = NULL;
		L->numitems--;
	} 
	else if (L->cursor == L->front)
	{
		deleteFront(L);
	} 
	else if (L->cursor == L->back)
	{
		deleteBack(L);		
	} 
	else if (L->cursor->prev == NULL && L->cursor->next == NULL) 
	{
		L->front = L->back = NULL;
		L->index--;
		L->numitems--;
	} 
	else
	{
		L->cursor->prev->next = L->cursor->next;
		L->cursor->next->prev = L->cursor->prev;
		freeNode(&L->cursor);
		L->cursor = NULL;
		L->index--;
		L->numitems--;
	}
}

// Other operations -----------------------------------------------------------
void printList(FILE* out, List L)
{
  	Node N;
  	
  	if (L == NULL)
  	{
      	printf("List Error: calling printList() on NULL List reference\n");
      	exit(1);
   	}
   	for (N = L->front; N != NULL; N = N->next)
   	{
        fprintf(out, "%d ", N->data);
   	}
}

List copyList(List L)
{
	List b = newList();

	if (L == NULL) 
	{
		printf("List Error: calling copyList() on NULL List reference\n");
		exit(1);
	}
	if (L != NULL)
	{
		for (Node n = L->front; n!=NULL; n = n->next)
		{
			append(b, n->data);
		}
	}
	return b;
}



































